support non-disturbing remote-instance-weight-loader

amysaq2023 opened the pull request:
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.ai to discuss further. -->

## Motivation

In https://github.com/sgl-project/sglang/pull/8215, SGLang has already supported a new load format: remote_instance, that allows new instance to load weights from another running instance. This approach can greatly improve weight loading time during instance initialization. However, since it use torch.distributed with NCCL as backend, it will disturb on-going inference requests: torch.distributed will always launch CUDA kernels for transferring weight tensors.

We come up with another backend option: TransferEngine, which will not disturbing any GPU workload and still, using RDMA to transfer weights.

<!-- Describe the purpose and goals of this pull request. -->

## Modifications

We initialize one TransferEngine for each ModelRunner and will register its weights to RDMA channel during initialization.

When initializing a new instance who wants to use `remote_instance` load_format with TransferEngine backend:
1. It  will send an HTTP request to retrieve the source instance's TransferEngine metadata, including RDMA keys mapped to the corresponding GPU memory addresses.
2. Using these RDMA keys, the new instance directly loads weights from the source's GPU memory.

How to use:
`python -m sglang.launch_server [args] \`
`  --load-format remote_instance \`
`  --remote-instance-weight-loader-seed-instance-ip [seed_instance_ip] \`
`  --remote-instance-weight-loader-seed-instance-service-port [seed_instance_service_port] \`
`  --remote-instance-weight-loader-backend "transfer_engine"`

<!-- Detail the changes made in this pull request. -->

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

amysaq2023 committed:
support non-disturbing remote-instance-weight-loader

This commit supports non-disturbing mode for loading weights from
remote instances.
Previouly, SGLang has already supported loading weights from remote
instances using torch.dist with NCCL as backend; however, this approach
will disturb on-going inference requests since torch.dist will launch
CUDA kernels to transfer weight tensors.
In this commit, it introduces another backend option: TransferEngine,
which can transfer weight tensors through RDMA without disturbing any
on-going GPU workload.

Signed-off-by: Anqi Shen <amy.saq@antgroup.com>

amysaq2023 commented:
cc @tianyuzhou95 @zhaochenyang20

stmatengss commented on python/sglang/srt/model_executor/model_runner.py:
```diff
@@ -519,6 +533,39 @@ def initialize(self, min_per_gpu_memory: float):
+            local_ip, "P2PHANDSHAKE", "rdma", ""  # auto discovery if empty
+        )
+        self.remote_instance_transfer_engine_session_id = (
+            f"{local_ip}:{self.remote_instance_transfer_engine.get_rpc_port()}"
```
Could you add a new environment variable or reuse an existing one for user-specific IB devices (e.g., MOONCAKE_DEVICE='ib0, ib1')?

amysaq2023 commented on python/sglang/srt/model_loader/loader.py:
```diff
@@ -1527,26 +1532,69 @@ def load_model(
+            # register memory region
+            weight_mr_dict = {}
+            for name, weight in model.named_parameters():
+                ret = load_config.remote_instance_weight_loader_transfer_engine.register_memory(
```
We actually found that using `batch_register_memory` and `register_memory` has the similar performance. The bottleneck seems to be the lock the ib driver trying to get to register memory during ioctl call.

zhaochenyang20 commented on python/sglang/srt/model_loader/loader.py:
```diff
@@ -1597,6 +1645,63 @@ def load_model_from_remote_instance(
         )
         torch.cuda.empty_cache()
 
+    def update_weights_from_remote_instance_by_transfer_engine(
```
As I said, use `load_model_xxx`

zhaochenyang20 reviewed (changes requested):
You are my hero! As solid as always. Sorry for waiting so long for the review. I shall also connect Mooncake team and MicroSoft AI for review.

JD-ETH commented on python/sglang/srt/model_loader/loader.py:
```diff
@@ -1527,26 +1532,69 @@ def load_model(
+                )
+            # register memory region
+            weight_mr_dict = {}
+            for name, weight in model.named_parameters():
```
curious, if we load model weights from the source again, does the transfer engine skip the expansive registration?

JD-ETH commented:
super cool! Just for my understanding, the `batch_transfer_sync_read` call is a single-sided write operation, and does not require additional TransferEngine instance at the target entity, correct?

amysaq2023 committed:
optimize registering memory region performance for remote instance loader

This commit reduces the time cost on registering memory regions for
using TransferEngine as backend to loading weights from remote instance,
by merging continuous memory blocks into one memory region.

Signed-off-by: Anqi Shen <amy.saq@antgroup.com>

amysaq2023 committed:
add unit tests for remote_instance_loader with TransferEngine backend

Signed-off-by: Anqi Shen <amy.saq@antgroup.com>

amysaq2023 commented:
> > > @amysaq2023 Hi, Can you share some perf data about `remote_instance with Transfer_engine`. I tried `remote_instance with nccl` with `deepseek-v3`(645GB), it synced within 3s on h200*8 with 8 NIC. Meantime, The on-going inference tasks would be blocked. And there are another question. How should we allocate the NICs for `GPU collective operations`, `KVcache transfer of PD disaggregation` and `remote_instance with TE`? In Our practices (9 NICs)， we use 8 NICs(assume ib0-7) pinned with GPUs for `GPU collective operations`, And 1 NIC(assume ib8) for `KVcache transfer of PD disaggregation`. And there is no NIC for `remote_instance`. if we allocate one of `ib0-7`, it would disturbe inference tasks; if allocate ib8 for `remote_instance`, it also disturbe KVCache Transfer, which also disturbe inference tasks in PD disaggregation. So how should we allocate NICs for `remote_instance` to minimize disturbing the on-going inference tasks. default allocate all NICs, let's MoonCake or System process it? And Can we limit the bandwidth of `remote_instance` , such as only use 1/4, 1/2 bandwidth?
> > 
> > 
> > Performance tested with 8*H20 on loading `Deepseek-V3`:
> > 
> > * register memory time: ~22s (~7s after WIP optimizing)
> > * transfer weights time: ~3s
> 
> @amysaq2023 thanks for your infos. Can you tell how to config `MOONCAKE_DEVICE` to reproduce the perf ? I tried this pr in our envs.
> 
> ```shell
> hardwares: 2 node of h200*8， with 8 * 400Gb/s NICs(mlx5_1~mlx5_8), 1 * 100Gb/s NIC(mlx5_0)
> softwars: mooncake-transfer-engine                 0.3.7.post2
> base envs:  export MOONCAKE_PROTOCOL="rdma"
> ```
> 
> and tested it with 3 settings, And its' perf as belows. Looking forward to your reply, thanks.
> 
> ```shell
> 1. don't set MOONCAKE_DEVICE, register_memory cost: 70.6599s, batch_transfer_sync_read: 6.3640s
> 2. export MOONCAKE_DEVICE='mlx5_0', register_memory cost: 8.7188s, batch_transfer_sync_read: 48.9649s
> 3. export MOONCAKE_DEVICE='mlx5_1,mlx5_2,mlx5_3,mlx5_4,mlx5_5,mlx5_6,mlx5_7,mlx5_8', register_memory cost: 60.5853s, batch_transfer_sync_read: 6.1010s
> ```
> 
> **nvidia-smi topo -m as belows**
> 
> ```shell
>         GPU0    GPU1    GPU2    GPU3    GPU4    GPU5    GPU6    GPU7    NIC0    NIC1    NIC2    NIC3    NIC4    NIC5    NIC6    NIC7    NIC8    CPU Affinity    NUMA Affinity   GPU NUMA ID
> GPU0     X      NV18    NV18    NV18    NV18    NV18    NV18    NV18    SYS     PIX     PHB     PHB     PHB     SYS     SYS     SYS     SYS     0-89    0               N/A
> GPU1    NV18     X      NV18    NV18    NV18    NV18    NV18    NV18    SYS     PHB     PIX     PHB     PHB     SYS     SYS     SYS     SYS     0-89    0               N/A
> GPU2    NV18    NV18     X      NV18    NV18    NV18    NV18    NV18    SYS     PHB     PHB     PIX     PHB     SYS     SYS     SYS     SYS     0-89    0               N/A
> GPU3    NV18    NV18    NV18     X      NV18    NV18    NV18    NV18    SYS     PHB     PHB     PHB     PIX     SYS     SYS     SYS     SYS     0-89    0               N/A
> GPU4    NV18    NV18    NV18    NV18     X      NV18    NV18    NV18    SYS     SYS     SYS     SYS     SYS     PIX     PHB     PHB     PHB     90-179  1               N/A
> GPU5    NV18    NV18    NV18    NV18    NV18     X      NV18    NV18    SYS     SYS     SYS     SYS     SYS     PHB     PIX     PHB     PHB     90-179  1               N/A
> GPU6    NV18    NV18    NV18    NV18    NV18    NV18     X      NV18    SYS     SYS     SYS     SYS     SYS     PHB     PHB     PIX     PHB     90-179  1               N/A
> GPU7    NV18    NV18    NV18    NV18    NV18    NV18    NV18     X      SYS     SYS     SYS     SYS     SYS     PHB     PHB     PHB     PIX     90-179  1               N/A
> NIC0    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS      X      SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS
> NIC1    PIX     PHB     PHB     PHB     SYS     SYS     SYS     SYS     SYS      X      PHB     PHB     PHB     SYS     SYS     SYS     SYS
> NIC2    PHB     PIX     PHB     PHB     SYS     SYS     SYS     SYS     SYS     PHB      X      PHB     PHB     SYS     SYS     SYS     SYS
> NIC3    PHB     PHB     PIX     PHB     SYS     SYS     SYS     SYS     SYS     PHB     PHB      X      PHB     SYS     SYS     SYS     SYS
> NIC4    PHB     PHB     PHB     PIX     SYS     SYS     SYS     SYS     SYS     PHB     PHB     PHB      X      SYS     SYS     SYS     SYS
> NIC5    SYS     SYS     SYS     SYS     PIX     PHB     PHB     PHB     SYS     SYS     SYS     SYS     SYS      X      PHB     PHB     PHB
> NIC6    SYS     SYS     SYS     SYS     PHB     PIX     PHB     PHB     SYS     SYS     SYS     SYS     SYS     PHB      X      PHB     PHB
> NIC7    SYS     SYS     SYS     SYS     PHB     PHB     PIX     PHB     SYS     SYS     SYS     SYS     SYS     PHB     PHB      X      PHB
> NIC8    SYS     SYS     SYS     SYS     PHB     PHB     PHB     PIX     SYS     SYS     SYS     SYS     SYS     PHB     PHB     PHB      X 
> 
> Legend:
> 
>   X    = Self
>   SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
>   NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
>   PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
>   PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
>   PIX  = Connection traversing at most a single PCIe bridge
>   NV#  = Connection traversing a bonded set of # NVLinks
> 
> NIC Legend:
> 
>   NIC0: mlx5_0
>   NIC1: mlx5_1
>   NIC2: mlx5_2
>   NIC3: mlx5_3
>   NIC4: mlx5_4
>   NIC5: mlx5_5
>   NIC6: mlx5_6
>   NIC7: mlx5_7
>   NIC8: mlx5_8
> ```

Could you please provide more information about how you set-up the test such as whether running SGLang instance in container and which SGLang image is used? Thanks.

amysaq2023 commented:
> super cool! Just for my understanding, the `batch_transfer_sync_read` call is a single-sided write operation, and does not require additional TransferEngine instance at the target entity, correct?

Actually the seed instance also needs to start TransferEngine server since it needs to register its RDMA memory regions for weights.

stmatengss commented on python/sglang/srt/managers/scheduler.py:
```diff
@@ -2659,11 +2662,18 @@ def run_scheduler_process(
                 "max_req_input_len": scheduler.max_req_input_len,
+                "tp_rank": tp_rank,
+                "remote_instance_transfer_engine_session_id": remote_instance_transfer_engine_session_id,
+                "remote_instance_transfer_engine_weights_info_dict": remote_instance_transfer_engine_weights_info_dict,
```
Check the global argument `remote_instance_weight_loader_support_transfer_engine` to determine whether to swap the info dict and session ID. These metadata are only needed for remote instances.

amysaq2023 replied:
Thanks for advice. Updated.

stmatengss commented on python/sglang/srt/model_loader/remote_instance_weight_loader_utils.py:
```diff
@@ -67,3 +75,109 @@ def trigger_transferring_weights_request(
+
+    end_tic = time.time()
+    logger.debug(f"Register memory region time: {(end_tic - start_tic):.4f}s")
+    return weight_mr_dict
```
If this method is unnecessary for memory registration, remove it or mark it as deprecated.

stmatengss commented on python/sglang/srt/server_args.py:
```diff
@@ -2043,6 +2068,20 @@ def _handle_other_validations(self):
+            logger.warning(
+                f"Failed to import mooncake.engine. Does not support using TransferEngine as remote instance weight loader backend."
+            )
+            self.remote_instance_weight_loader_support_transfer_engine = False
```
Omit this comment. Use importlib should be better.

zhaochenyang20 reviewed (changes requested):
One final request, and I think this should also be kept in the docs of SGLang:

Adding documents to https://docs.sglang.io/advanced_features/

You can follow what others' docs and ping me to review the docs.

I think we should link the docs to RFork's blog also.

amysaq2023 committed:
add document for R-Fork

Signed-off-by: Anqi Shen <amy.saq@antgroup.com>

amysaq2023 commented:
> One final request, and I think this should also be kept in the docs of SGLang:
> 
> Adding documents to https://docs.sglang.io/advanced_features/
> 
> You can follow what others' docs and ping me to review the docs.
> 
> I think we should link the docs to RFork's blog also.

Thanks for advice :) We have added the document for R-Fork in docs/advanced_features/rfork.md

amysaq2023 commented:
> I see some tests of remote-instance-weight-loader cannot pass.
> 
> ```
> FAILED (errors=1)
> [CI Test Method] TestLoadWeightsFromRemoteInstance.test_load_weights_from_remote_instance
> .
> .
> End (6/7):
> filename='/public_sglang_ci/runner-l1d-gpu-67/_work/sglang/sglang/test/srt/test_load_weights_from_remote_instance.py', elapsed=8, estimated_time=72
> .
> .
> 
> 
> ✗ FAILED: test_load_weights_from_remote_instance.py returned exit code 1
> 
> Fail. Time elapsed: 1083.95s
> 
> ============================================================
> Test Summary: 6/8 passed
> ============================================================
> ✓ PASSED:
>   test_disaggregation_basic.py
>   hicache/test_hicache_storage_3fs_backend.py
>   hicache/test_hicache_storage_file_backend.py
>   ep/test_moe_ep.py
>   rl/test_update_weights_from_distributed.py
>   models/test_kimi_linear_models.py
> 
> ✗ FAILED:
>   test_load_weights_from_remote_instance.py (exit code 1)
> ============================================================
> ```

Thanks for pointing out. Test case issue fixed.

stmatengss reviewed (approved):
LGTM. Only one concern: for the CI matrix in test_load_weights_from_remote_instance, should we account for the total runtime?

zhaochenyang20 commented on python/sglang/srt/model_loader/loader.py:
```diff
@@ -1992,26 +1534,59 @@ def load_model(
+                register_memory_region_v2(
+                    model, load_config.remote_instance_weight_loader_transfer_engine
+                )
+            )
```
As mentioned in the blog, since the register memory region shall takes several seconds, could we add a log here:

```python
logger.info("Initializing TransferEngine and registering memory regions (this may take a few seconds)...")
# ... registration code ...
logger.info("TransferEngine initialized successfully.")
```

zhaochenyang20 commented on python/sglang/srt/model_loader/loader.py:
```diff
@@ -1992,26 +1534,59 @@ def load_model(
+            if not success:
+                raise RuntimeError(
+                    "Failed to load weights from remote instance via transfer engine."
+                )
```
Should we have a timeout here?

Kangyan-Zhou committed:
Merge branch 'main' into amy/non-disturbing-remote-instance-weight-loader

amysaq2023 commented on python/sglang/srt/entrypoints/engine.py:
```diff
@@ -926,9 +932,24 @@ def _launch_subprocesses(
+        template_manager,
+        scheduler_info,
+        port_args,
+        remote_instance_transfer_engine_info,
```
The problem here is, scheduler_info has an assumption that all schedulers have the same information, and only return scheduler_infos[0], which is not true for remote instance transfer engine info. Each scheduler has different transfer engine information here.

zhyncs commented:
This PR broke the B200 CI. We need to be more careful during review and before merging PRs — the main branch has been breaking too frequently lately.
@amysaq2023 @zhaochenyang20 @Kangyan-Zhou

amysaq2023 commented:
> > @amysaq2023 It conflicts with some refactors here #14869. Would be great if you can keep the interface of `_launch_subprocesses` unchanged.
> 
> Sure. Working on refactoring this.

PR addressed above comments: https://github.com/sgl-project/sglang/pull/14971
@merrymercy @zhaochenyang20
